src/i18n.tsx

```diff
@@ -1,34 +1,43 @@
 import React from 'react'
 import { i18n } from '@lingui/core'
 import { I18nProvider } from '@lingui/react'
-import { detect, fromUrl } from '@lingui/detect-locale'
-import { ReactNode, useEffect } from 'react'
+import { detect, fromNavigator } from '@lingui/detect-locale'
+import { ReactNode } from 'react'
 
-export const locales = ['en', 'pseudo-en', 'de', 'es-AR', 'es-US', 'it-IT', 'iw', 'ro', 'ru', 'vi', 'zh-CN', 'zh-TW']
-export const defaultLocale = 'en'
+export const supportedLocales = [
+  'en',
+  'pseudo-en',
+  'de',
+  'es-AR',
+  'es-US',
+  'it-IT',
+  'iw',
+  'ro',
+  'ru',
+  'vi',
+  'zh-CN',
+  'zh-TW',
+]
+export const defaultFallback = () => 'en'
 
-const getDetectedLocale = () => {
-  const detected =
-    detect(
-      fromUrl('lang'), // helps local development
-      defaultLocale
-    ) ?? defaultLocale
-  return locales.includes(detected) ? detected : defaultLocale
+export const isSupportedLocale = (locale: string): boolean => supportedLocales.includes(locale)
+
+export const getDetectedLocale = (): string => {
+  const detected = detect(fromNavigator())
+  return detected && isSupportedLocale(detected) ? detected : defaultFallback()
 }
 
 export async function dynamicActivate(locale: string) {
-  const { messages } = await import(`@lingui/loader!./locales/${locale}.po`)
-  i18n.loadLocaleData(locale, { plurals: () => null })
-  i18n.load(locale, messages)
-  i18n.activate(locale)
+  try {
+    const { messages } = await import(`@lingui/loader!./locales/${locale}.po`)
+    i18n.loadLocaleData(locale, { plurals: () => null })
+    i18n.load(locale, messages)
+    i18n.activate(locale)
+  } catch (error) {
+    console.error(`Failed to load locale data for ${locale}`, error)
+  }
 }
 
 export function LanguageProvider({ children }: { children: ReactNode }) {
-  useEffect(() => {
```

Comment:
i think it makes more sense for the dynamic activate effect to live here, and you just compute the locale to use from the state in the redux store + navigator + url in the LanguageProvider component

Reply:
my rationale is that
a) it keeps the redux state folder separate from user agent/browser stuff
b) this component can get access to the url and user agent because it is intended to be rendered in the dom